include/clad/Differentiator/NumericalDiff.h

```diff
@@ -0,0 +1,418 @@
+      return static_cast<T*>(ptr);
+    }
+
+    /// A function to free the space previously allocated.
+    void free_buffer() {
+      while (data.size()) {
+        U ptr = data.back();
+        data.pop_back();
+        free(ptr);
+      }
+    }
+  };
+
+  /// The precison to do the numerical differentiation calculations in.
+  using precision = double;
+
+  /// A function to make sure the step size being used is machine representable.
+  /// It is likely that if do not have a similar construct, we may end up with
+  /// catastrophic cancellations, hence resulting into a 0 derivative over a
+  /// large interval.
+  ///
+  /// \param[in] x Input to the target function.
+  /// \param[in] h The h to make representable.
+  ///
+  /// \returns A value of h that does not result in catastrohic cancellation.
+  precision make_h_representable(precision x, precision h) {
+    precision xph = x + h;
+    precision dx = xph - x;
+
+    // If x+h ~ x, we should look for the next closest value so that (x+h) - x
+    // != 0
+    if (dx == 0)
+      h = std::nextafter(x, std::numeric_limits<precision>::max()) - x;
+
+    return h;
+  }
+
+  /// A function to return the h value to use.
+  ///
+  /// \param[in] arg The input argument to adjust and get h for.
+  ///
+  /// \returns A calculated and adjusted h value.
+  precision get_h(precision arg) {
+    // First get a sutiable h value, we do all of this in elevated precision
+    // (default double). Maximum error in h = eps^4/5
+    precision h = std::pow(11.25 * std::numeric_limits<precision>::epsilon(),
+                           (double)1.0 / 5.0);
+    h = make_h_representable(arg, h);
+    return h;
+  }
+
+  /// A function to update scalar parameter values given a multiplier and the
+  /// argument value. Custom data types can be supported as input in the case
+  /// that this function is overloaded.
+  ///
+  /// This function should be overloaded as follows:
+  /// \code
+  /// template <typename <--your-data-type-->, typename U>
+  /// <--your-data-type-->* updateIndexParamValue(<--your-data-type--> arg, 
+  ///                           std::size_t idx, std::size_t currIdx,
+  ///                          int multiplier, precision& h_val,
+  ///                          ManageBufferSpace<U>& bufferManager,
+  ///                          std::size_t n = 0, std::size_t i = 0) {
+  ///   if (idx == currIdx) {
+  ///     // Add your type specific code here that would essentially translate
+  ///     // to `return arg + multiplier * h` (h is some very small value)   
+  ///     h_val = //Save the 'precision' equivalent value of h here, not doing  
+  ///     // so will lead to undefined behaviour.
+  ///   }
+  ///   return arg;
+  /// }
+  /// \endcode
+  ///
+  /// \param[in] arg The argument to update.
+  /// \param[in] idx The index of the current parameter in the target function
+  /// parameter pack.
+  /// \param[in] currIdx The index value of the parameter to be
```

Comment:
```suggestion
  /// \param[in] currIdx - The index value of the parameter to be
```
I usually use the dash to make the name more clear. Not sure if that is an issue in doxygen generated html, but definitely helps reading the code.